### PR TITLE
Complete the Regex building code

### PR DESCRIPTION
**Pull request by mominul**
The first commit adds a huge number of tests that were failing because of some unimplemented aspects of Regex string building.
The second commit implements the specific features of the Regex building code to fix the failing tests.

Need to run benchmarks to check if this introduces regressions in the Phonetic conversion code.

Thanks!

### CODE REVIEW

File: src/parser.rs [hunk 1]
```diff
@@ -99,6 +106,7 @@ impl Phonetic {
                             Some(rule) => output.push_str(rule.replace_with),
                             None => output.push_str(default_replacement),
                         };
+                        output.push_str(extra);
```
Comment (gulshan) [inline src/parser.rs:109]: As this push is being done for both if and else branches, it can be outside branching.

**Comment by gulshan** [comment]
Should we go for a more generic name like `Matcher` instead of `Phonetic` for our public struct?

**Comment by gulshan** [comment]
> Should we go for a more generic name like `Matcher` instead of `Phonetic` for our public struct?

Or just `Parser`, same as the file/module name, following Rust convention?

**Comment by mominul** [comment]
I have addressed your review, thanks!
> Or just Parser, same as the file/module name, following Rust convention?
`Parser` seems to be conventional :+1: 
If we do the naming change, how about changing the name of `regex_patterns.rs` to `regex_parser.rs` of simply `regex.rs`?

File: src/regex_patterns.rs [hunk 2]
```diff
@@ -187,4 +189,15 @@ impl crate::parser::Phonetic {
...
     }
+
+    pub fn convert_regex(&self, raw_input: &str) -> String {
+        let input: String = raw_input.to_ascii_lowercase().chars().filter(|&ch| !IGNORE.contains(ch)).collect();
+        let mut output = String::with_capacity(input.len() * 60);
+        output.push('^');
+
+        self.convert_internal(&input, &mut output, "(্[যবম])?(্?)([ঃঁ]?)");
+
+        output.push('$');
```
Comment (gulshan) [inline src/regex_patterns.rs:200]: Small styling opinion. I think the empty lines should be before and after push here, not between them. 
```suggestion

        output.push('^');
        self.convert_internal(&input, &mut output, "(্[যবম])?(্?)([ঃঁ]?)");
        output.push('$');

```

**Comment by gulshan** [comment]
I have just pushed the renaming commit- "Phonetic" to "Parser". Please rebase.
Regarding the name of the regex file, it's up to you. Choose whichever makes more sense.

**Comment by mominul** [comment]
Benchmarks:
<pre>group                    after                                  before
-----                    -----                                  ------
okkhor ami               1.02    269.9±3.36ns        ? ?/sec<font color="#26A269"><b>    1.00    265.0±2.42ns        ? ?/sec</b></font>
okkhor bistarito         1.03    698.0±8.70ns        ? ?/sec<font color="#26A269"><b>    1.00    679.7±7.87ns        ? ?/sec</b></font>
okkhor kormo             1.02    420.0±7.14ns        ? ?/sec<font color="#26A269"><b>    1.00    411.6±5.12ns        ? ?/sec</b></font>
okkhor long word         1.03   548.8±10.35ns        ? ?/sec<font color="#26A269"><b>    1.00    532.8±7.03ns        ? ?/sec</b></font>
okkhor new               1.00     22.5±0.31µs        ? ?/sec<font color="#26A269"><b>    1.00     22.4±0.30µs        ? ?/sec</b></font>
okkhor sentence 1        1.02  1527.6±15.06ns        ? ?/sec<font color="#26A269"><b>    1.00  1505.0±17.88ns        ? ?/sec</b></font>
okkhor sentence 2        1.01      5.2±0.04µs        ? ?/sec<font color="#26A269"><b>    1.00      5.1±0.06µs        ? ?/sec</b></font>
okkhor sonar bangla      1.02  1369.0±13.02ns        ? ?/sec<font color="#26A269"><b>    1.00  1342.0±12.26ns        ? ?/sec</b></font>
</pre>

**Pull request merged**